hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/impl/TestBlockManagerImpl.java

```diff
@@ -196,6 +200,72 @@ public void testPutAndGetBlock(ContainerTestVersionInfo versionInfo)
+      assertEquals(1, db.getStore().getMetadataTable().get(containerData.getBlockCountKey()));
+
+      // 2. Put Block with bcsId = 2, Overwrite BCS ID = false
+      BlockData blockData2 = createBlockData(1L, 3L, 1, 0, 2048, 2);
+      blockManager.putBlockForClosedContainer(keyValueContainer, blockData2, false);
+
+      // The block should be written, but we won't be able to read it, As expected BcsId < container's BcsId
+      // fails during block read.
+      assertThrows(StorageContainerException.class, () -> blockManager
+          .getBlock(keyValueContainer, blockData2.getBlockID()));
+      fromGetBlockData = db.getStore().getBlockDataTable().get(containerData.getBlockKey(blockData2.getLocalID()));
+      assertEquals(2, containerData.getBlockCount());
+      // BcsId should still be 1, as the BcsId is not overwritten
+      assertEquals(1, containerData.getBlockCommitSequenceId());
+      assertEquals(2, fromGetBlockData.getBlockCommitSequenceId());
+      assertEquals(2, db.getStore().getMetadataTable().get(containerData.getBlockCountKey()));
+      assertEquals(1, db.getStore().getMetadataTable().get(containerData.getBcsIdKey()));
+
+      // 3. Put Block with bcsId = 2, Overwrite BCS ID = true
+      // This should succeed as we are overwriting the BcsId, The container BcsId should be updated to 2
+      // The block count should not change.
+      blockManager.putBlockForClosedContainer(keyValueContainer, blockData2, true);
+      fromGetBlockData = blockManager.getBlock(keyValueContainer, blockData2.getBlockID());
+      assertEquals(2, containerData.getBlockCount());
+      assertEquals(2, containerData.getBlockCommitSequenceId());
+      assertEquals(2, fromGetBlockData.getBlockCommitSequenceId());
+      assertEquals(2, db.getStore().getMetadataTable().get(containerData.getBlockCountKey()));
+      assertEquals(2, db.getStore().getMetadataTable().get(containerData.getBcsIdKey()));
+
+      // 4. Put Block with bcsId = 1 < container bcsId, Overwrite BCS ID = true
+      // We are overwriting an existing block with lower bcsId than container bcdId. Container bcsId should not change
+      BlockData blockData3 = createBlockData(1L, 1L, 1, 0, 2048, 1);
+      blockManager.putBlockForClosedContainer(keyValueContainer, blockData3, true);
+      fromGetBlockData = blockManager.getBlock(keyValueContainer, blockData3.getBlockID());
+      assertEquals(3, containerData.getBlockCount());
+      assertEquals(2, containerData.getBlockCommitSequenceId());
+      assertEquals(1, fromGetBlockData.getBlockCommitSequenceId());
+      assertEquals(3, db.getStore().getMetadataTable().get(containerData.getBlockCountKey()));
+      assertEquals(2, db.getStore().getMetadataTable().get(containerData.getBcsIdKey()));
+
+      // Used Bytes is only update by writeChunk, so it should be 0 here.
+      assertEquals(0, db.getStore().getMetadataTable().get(containerData.getBytesUsedKey()));
+    }
```

Comment:
Is this right? The used bytes in memory is updated by write chunk, but put block should flush the in-memory value to the DB.

Reply:
Updated the comment.